Move EventJournalConfig inside MapConfig and CacheConfig

mmedenjak opened the pull request:
Initially, event journal configuration was put as a top-level config
because we are not able to change CacheConfig because of client
backwards compatibility (see BinaryInterface annotation).
Now with 4.0, we will remove this restriction which means event journal
config can be part of both map and cache config.
Using it this way will eliminate some of the confusion as up until now
the user had to additionally specify the mapName and cacheName on the
event journal configuration to 'connect' the event journal configuration
to the map configuration.

Remaining:
- change client protocol to remove merkle tree codec and append the map codec

Depends on: https://github.com/hazelcast/hazelcast/pull/15180
EE: https://github.com/hazelcast/hazelcast-enterprise/pull/3047
Client protocol PR: https://github.com/hazelcast/hazelcast-client-protocol/pull/190

vbekiaris commented on hazelcast/src/main/java/com/hazelcast/config/MapConfig.java:
```diff
@@ -993,6 +1040,12 @@ public final boolean equals(Object o) {
         if (quorumName != null ? !quorumName.equals(that.quorumName) : that.quorumName != null) {
             return false;
         }
+        if (merkleTreeConfig != null ? !merkleTreeConfig.equals(that.merkleTreeConfig) : that.merkleTreeConfig != null) {
```
minor: if `merkleTreeConfig` and `eventJournalConfig` cannot be `null`, some null checks can be ommited for readability.

vbekiaris commented on hazelcast/src/main/java/com/hazelcast/config/MapConfig.java:
```diff
@@ -886,6 +893,46 @@ public MapConfig setHotRestartConfig(HotRestartConfig hotRestartConfig) {
+     * @param merkleTreeConfig merkle tree config
+     * @return this {@code MapConfig} instance
+     */
+    public MapConfig setMerkleTreeConfig(MerkleTreeConfig merkleTreeConfig) {
```
is a `null` argument allowed here?

mmedenjak replied:
Probably not. `null` will bring just problems somewhere down the line. I'll add a check.

vbekiaris commented on hazelcast/src/main/java/com/hazelcast/config/MerkleTreeConfig.java:
```diff
@@ -214,17 +182,14 @@ public boolean equals(Object o) {
-            return false;
-        }
-        return mapName != null ? mapName.equals(that.mapName) : that.mapName == null;
+        return depth == that.depth;
```
`MerkleTreeConfig` lacks a test for `equals`/`hashCode`, can you add one?

mmedenjak replied:
Will do!

vbekiaris commented on hazelcast/src/test/java/com/hazelcast/config/ConfigCompatibilityChecker.java:
```diff
@@ -223,12 +217,27 @@ private static boolean isCompatible(CollectionConfig c1, CollectionConfig c2) {
+        boolean c1Disabled = c1 == null || !c1.isEnabled();
+        boolean c2Disabled = c2 == null || !c2.isEnabled();
+        return c1 == c2 || (c1Disabled && c2Disabled) || (c1 != null && c2 != null
+                && nullSafeEqual(c1.getDepth(), c2.getDepth()));
```
minor: no reason to use `nullSafeEqual` for `int` comparison here (same below)

mmedenjak replied:
True, copy-paste error, thanks!

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/config/CacheConfig.java:
```diff
@@ -520,6 +522,7 @@ public void writeData(ObjectDataOutput out) throws IOException {
         out.writeBoolean(isStatisticsEnabled);
         out.writeBoolean(hotRestartConfig.isEnabled());
```
Sure, it looked a bit fishy, will change it with this PR since it's not a big deal.

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/config/AbstractCacheConfig.java:
```diff
@@ -454,6 +477,8 @@ public int hashCode() {
         result = 31 * result + (isStatisticsEnabled ? 1 : 0);
         result = 31 * result + (isStoreByValue ? 1 : 0);
         result = 31 * result + (isManagementEnabled ? 1 : 0);
+        result = 31 * result + (hotRestartConfig != null ? hotRestartConfig.hashCode() : 0);
```
True, will change that!

mmedenjak commented:
@vbekiaris thanks for the review! It may not be easy to see from the PR but I've actually separated the PRs for moving the MerkleTreeConfig and moving the EventJournalConfig since I thought it would be easier for the reviewers. Maybe I was wrong and it should have been one single PR. Nevermind now, I'll keep it as-is and I'll address the merkle-tree related comments in the merkle-tree config PR.

mmedenjak commented:
@vbekiaris Addressed review comments in the [separate PR commit](https://github.com/hazelcast/hazelcast/pull/15180/commits/4c4f2ee606d884eec2bd5c2516c89db14151237b), please take a look.

mmedenjak commented:
Thank you for the reviews, guys 🙇
